openedx/core/lib/graph_traversals.py

```diff
@@ -0,0 +1,257 @@
+"""
+This module contains generic generator functions for traversing tree
+(and DAG) structures.  It is agnostic to the underlying data structure
+and implementation of the tree object.  It does this through dependency
+injection of the tree's accessor functions: get_parents and
+get_children.
+
+Additionally, it supports the visitor design pattern by accepting a
+get_result function that is called on each visited node.
+
+The following depth-first traversal methods are implemented:
+
+* Pre-order: Parent yielded before children; child with multiple
+parents is yielded when first encountered.
+
+* Topological: Parent yielded before children; child with multiple
+parents yielded only after all its parents are visited.
+
+* Post-order: Children yielded before its parents.
```

Comment:
I think this module docstring is great. One small ask I have here is to give examples of when you would pick a particular ordering over another. I think it's easy for a developer to gloss over this without understanding why you would ever want to do, say, a post-order traversal.